### DISCUSSION TITLE
Unable to dockerize my NextJS app!

### DISCUSSION THREAD

**Discussion by schandillia** [vercel/next.js#61625] [Help]
### Summary

I am trying to dockerize my Next (14) app’s production build using the exact same file (pasted below) as given in this official repo: [https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile).

The only addition is that I have a couple of Prisma commands in the **builder** stage right after `COPY . .`.

But no matter how many times I try, `docker-compose up --build` keeps failing with the following error:

`Error: Cannot find module '/app/server.js'`

I also have the following entry in my **next.config.js**: `output: "standalone",`

Can anyone at Vercel please help figure out what I am missing?

### Additional information

```js
FROM node:20.10-alpine AS base

# Install dependencies only when needed
FROM base AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache bash
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json pnpm-lock.yaml* ./
RUN corepack enable pnpm
RUN pnpm i --frozen-lockfile


# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

# Next.js collects completely anonymous telemetry data about general usage.
# Learn more here: https://nextjs.org/telemetry
# Uncomment the following line in case you want to disable telemetry during the build.
ENV NEXT_TELEMETRY_DISABLED 1
# Generate Prisma Client
RUN npx prisma db push
RUN npx prisma generate

RUN corepack enable pnpm
RUN pnpm run build

# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000

ENV PORT 3000
# set hostname to localhost
ENV HOSTNAME "0.0.0.0"

# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD ["node", "server.js"]
```

### Example

https://github.com/schandillia/poco/tree/auth

**Comment by ludo237** [comment]
Try this in your build phase: `ENV NEXT_PRIVATE_STANDALONE true`

This fixed the issue for me although i don't know what is the cause. Seems like latest NextJs ignores the 'output' key

**Reply by RobertHH-IS** [reply]
This right here - standalone in next.config not enough

**Reply by carlos-talavera** [reply]
Sadly it didn't work for me. I'm using pnpm and locally it works fine. It just gives the error when trying to dockerize it. Apparently the files are not being generated, hence they're not being copied. It's weird.

**Reply by carlos-talavera** [reply]
Pretty odd, I used the Dockerfile of [this article](https://dev.to/vorillaz/how-to-dockerize-a-nextjs-app-4e4h), it worked and then I was trying to narrow the issue down on my original Dockerfile and tried changing the way I was fetching and installing dependencies using pnpm, but it wasn't that. Using or not using `ENV NEXT_PRIVATE_STANDALONE true` was the same, it'd still work. Then I was thinking about the group and the user, but both of them worked. Then I thought about the order of the environment variables, but it was the same result. No matter what I changed, it'd keep working. So I just tried with my original Dockerfile and magically worked. Pretty odd actually.

**Comment by Zoney** [comment]
Encountering the same thing :(

**Reply by Walfhand** [reply]
I also had the same problem, I solved it by renaming my workdir in my dockerfile, for some reason if I name my wordir '/app' it doesn't work, for example if you put /src for me it works

**Reply by she11fish** [reply]
> I also had the same problem, I solved it by renaming my workdir in my dockerfile, for some reason if I name my wordir '/app' it doesn't work, for example if you put /src for me it works

I tried it for myself when I got the issue. From my understanding of the situation. It seems like if you have `/app` as read only, then the build will not work.
When changing it to `/src`, you're fixing the permission issues as you have `/src`, `node_modules`, and `.next` working properly since they execute permissions for `server.js` to work.

You can add this before the line `USER nextjs` in case you want to fix the permission issues.
```
# assuming we're in the working directory /app 
RUN chmod +x . && chmod +x -R .next node_modules
```
And optionally strip out the unnecessary permissions on top of it.
```
RUN chmod -wx -R .
# the rest
```
Combining it together would look something like this
```
RUN chmod -R a-w+x . && chmod -R a+x .next node_modules
```

Please let me know if this works or helps anyone.

**Comment by mtansk** [comment]
I have experienced the same issue a while ago and, as far as I remember, I fixed it by replacing 
```
# set hostname to localhost
ENV HOSTNAME "0.0.0.0"

# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD ["node", "server.js"]
```

with 

`CMD HOSTNAME="0.0.0.0" node server.js`

**Comment by GeuntaBuwono** [comment]
if you got `distDir` on next.config.ts this needs to be removed
